src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp

```diff
@@ -219,6 +219,105 @@ ObjectStorageQueueMetadata::tryAcquireBucket(const Bucket & bucket, const Proces
     return ObjectStorageQueueOrderedFileMetadata::tryAcquireBucket(zookeeper_path, bucket, processor, log);
 }
 
+void ObjectStorageQueueMetadata::alterSettings(const SettingsChanges & changes)
+{
+    const fs::path alter_settings_lock_path = zookeeper_path / "alter_settings_lock";
+    zkutil::EphemeralNodeHolder::Ptr alter_settings_lock;
+    auto zookeeper = getZooKeeper();
+
+    /// We will retry taking alter_settings_lock for the duration of 5 seconds.
+    /// Do we need to add a setting for this?
+    const size_t num_tries = 100;
+    for (size_t i = 0; i < num_tries; ++i)
+    {
+        alter_settings_lock = zkutil::EphemeralNodeHolder::tryCreate(alter_settings_lock_path, *zookeeper, toString(getCurrentTime()));
+
+        if (alter_settings_lock)
+            break;
+
+        if (i == num_tries - 1)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to take alter setting lock");
+
+        sleepForMilliseconds(50);
+    }
+
+    Coordination::Stat stat;
+    auto metadata_str = zookeeper->get(fs::path(zookeeper_path) / "metadata", &stat);
+    auto metadata_from_zk = ObjectStorageQueueTableMetadata::parse(metadata_str);
+    auto new_table_metadata{table_metadata};
+
+    for (const auto & change : changes)
+    {
+        if (endsWith(change.name, "processing_threads_num"))
+        {
+            const auto value = change.value.safeGet<UInt64>();
+            if (table_metadata.processing_threads_num == value)
+            {
+                LOG_TRACE(log, "Setting `processing_threads_num` already equals {}. "
+                        "Will do nothing", value);
+                return;
```

Comment:
Because of this code in `AlterCommands.cpp` we always get the last value in duplicated settings https://github.com/ClickHouse/ClickHouse/blob/a6fff970ac702f2a093ff5397d781cff68a99d9a/src/Storages/AlterCommands.cpp#L838-L851

This is in `commands.apply(new_metadata, local_context);`.
I can add a check for duplicated settings just before this `apply` method, but this is not done for `MergeTreeSettings`, so may be we'd leave for s3queue same behaviour as merge tree settings have.

Reply:
I mean let's we have a storage

```
CREATE TABLE queue (...) ENGINE = S3Queue(...)
SETTINGS
    processing_threads_num=1,
    loading_retries=0;
```

I want to change `loading_retries` to `1`, then I execute `ALTER TABLE queue MODIFY SETTINGS processing_threads_num=1, loading_retries=1`. Then we will return from the functions and ignore the rest of the settings. Maybe you wanted to write `continue` instead of `return`?


```suggestion
                continue;
```